### ISSUE TITLE
[GroupBox] Missing

### ISSUE THREAD

**Issue by SuperJMN** [AvaloniaUI/Avalonia#823]
There is no `GroupBox` like In WPF. Although it's not essential, this control is useful to label groups of controls.

**TEMPORARY TRICK:**
Although it doesn't exist, it can be emulated with a `HeaderedContentControl` with a custom Style:

```
        <Style Selector="HeaderedContentControl">
            <Setter Property="Template">
                <Setter.Value>
                    <ControlTemplate>
                        <Border>
                            <StackPanel>
                                <ContentPresenter TextBlock.FontWeight="Bold" Content="{TemplateBinding Header}" />
                                <Border
                                    BorderBrush="{TemplateBinding Background}"
                                    BorderThickness="2"
                                    CornerRadius="5">
                                    <ContentPresenter Content="{TemplateBinding Content}" />
                                </Border>
                            </StackPanel>
                        </Border>
                    </ControlTemplate>
                </Setter.Value>
            </Setter>
        </Style>
```

**Comment by derekantrican** [comment]
Is this still being considered or is it old?

**Comment by FoggyFinder** [comment]
Probably the issue should be closed: #3913

**Comment by derekantrican** [comment]
Yeah, I think agreed. I also saw that PR and it seems the conclusion was it's not going to be added because "It is just a styled HeaderedContentControl"?

**Comment by Gillibald** [comment]
#3913

**Comment by derekantrican** [comment]
For people who look at this in the future (which I'm sure will be the case unless one day Avalonia officially implements `GroupBox`), the workaround by @SuperJMN is great - you just have to assign the `Border` property, which is different from how normal WPF works (I think it's just black by default):

```
        <Style Selector="HeaderedContentControl">
            <Setter Property="Template">
                <Setter.Value>
                    <ControlTemplate>
                        <Border>
                            <StackPanel>
                                <ContentPresenter TextBlock.FontWeight="Bold" Content="{TemplateBinding Header}" />
                                <Border
                                    BorderBrush="{TemplateBinding Background}"
                                    BorderThickness="2"
                                    CornerRadius="5">
                                    <ContentPresenter Content="{TemplateBinding Content}" />
                                </Border>
                            </StackPanel>
                        </Border>
                    </ControlTemplate>
                </Setter.Value>
            </Setter>
        </Style>
        <HeaderedContentControl Header="Header" Background="Black">
          <TextBlock Text="text" Margin="5"/>
        </HeaderedContentControl>
```

**Comment by BinToss** [comment]
@derekantrican's examples are great, but the Classic style in that example will not adapt to Theme or Style changes i.e. Header background will remain solid white when the "real" background changes.
I've considered two solutions to this. One utlizes the Border's Clip property. The other utilizes the Border's OpacityMask. I'm attempting the former, but I'm struggling to make the Clip geometry recalculate when the source Bounds change. Both will likely require classes for conversions and/or observers.

## `Border.Clip = BoundsA - BoundsB`
<br/>
<details>
<summary>
<b>(Click to see images)</b> At one point, I was excluding <code>GBHeader.Bounds</code> (GBHeader being the Border containing the text block in the original example) from <code>GBContent.Bounds</code>, but an unexpectedly large part of the latter's Border is clipped despite the resulting geometry not affecting that region according the Dev Tools.
</summary>
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1121703098612731934/image.png" />
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1120640813077119017/image.png" />
</details>
<details><summary><b>(Click to see images)</b> However, I now exclude <code>GBHeaderText.Bounds</code> from <code>GBGrid.Bounds</code> to avoid that particular issue.</summary>
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1122408984083705946/image.png" />
</details>

**This implementation currently has two issues.**
1. The `CombinedGeometry` assigned to `GBContent.Clip` is not recalculated when the `Rect`-value Bounds it was created from have changed.
2. The `CombinedGeomtry` assigned to `GBContent.Clip` partially obscures the presented contents. I believe this could be remedied by getting the `TransformBounds`, but the lack of this property in my Avalonia 11.0.0-rc1.1 application's Dev Tools window indicates it no longer exists or may be inaccessible under certain conditions. The suspected issue is that two Bounds have different origin points relative to the window, so excluding one Bounds from the other will result in one being offset from its expected position in the CombinedGeometry

<details><summary><h4>GroupBoxClassic.axaml</h4></summary>

```xml
<Styles
	xmlns="https://github.com/avaloniaui"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="https://github.com/avaloniaui ../../../AvaloniaSchema-11.0.0.xsd"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	x:CompileBindings="True"
	xmlns:conv="clr-namespace:HXE.GUI.Converters;assembly=HXE">

	<!-- AvaloniaSchema is generated by Avant Garde -->

	<Style
		Selector="HeaderedContentControl">
		<SimpleTheme />
		<Style.Resources>
			<conv:RectGeometryConverter x:Key="rectConv" />
		</Style.Resources>
		<Setter
			Property="Template">
			<Setter.Value>
				<ControlTemplate>
					<Grid Name="GBGrid">
						<Grid.RowDefinitions>
							<RowDefinition Height="Auto" />
							<RowDefinition Height="*" />
						</Grid.RowDefinitions>
						<Grid.ColumnDefinitions>
							<ColumnDefinition Width="Auto" />
							<ColumnDefinition Width="*" />
						</Grid.ColumnDefinitions>


						<TextBlock Name="GBHeaderText"
							Margin="5,0,0,0"
							Padding="5,0,5,0"
							ZIndex="1"
							TextBlock.FontWeight="Bold" Text="{TemplateBinding Header}">
							<TextBlock.Bounds />
						</TextBlock>

						<Border Name="GBContent"
							Grid.RowSpan="2" Grid.ColumnSpan="2"
							Margin="0,10,0,0"
							Padding="5,10,5,5"
							BorderBrush="{TemplateBinding Foreground}"
							BorderThickness="1"
							Clip="{Binding}">
							<Border.Clip>
								<CombinedGeometry x:DataType="Rect"
									GeometryCombineMode="Exclude"
									Geometry1="{Binding ElementName=GBGrid, Path=Bounds, Mode=OneWay, Converter={StaticResource rectConv}}"
									Geometry2="{Binding ElementName=GBHeaderText, Path=Bounds, Mode=OneWay, Converter={StaticResource rectConv}}">
								</CombinedGeometry>
							</Border.Clip>
							<ContentPresenter Content="{TemplateBinding Content}" />
						</Border>
					</Grid>
				</ControlTemplate>
			</Setter.Value>
		</Setter>
	</Style>
</Styles>

```
</details>
<details><summary><h4>RectGeometryConverter.cs</h4></summary>

```cs
using System;
using System.Globalization;
using Avalonia;
using Avalonia.Data;
using Avalonia.Data.Converters;
using Avalonia.Media;

namespace HXE.GUI.Converters;

public class RectGeometryConverter : IValueConverter
{
    public static readonly RectGeometryConverter Instance = new();

    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is Rect rect)
            return new RectangleGeometry(rect);

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is RectangleGeometry rectangle)
            return rectangle.Rect;

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

}

```
</details>

**Comment by yll690** [comment]
@BinToss Thank you for your styles. It's very useful. And I found replacing the CombinedGeometry with a MultiBinding solves the first known issue.
```
<Border.Clip>
  <MultiBinding Converter="{StaticResource rectConv}">
    <Binding ElementName="GBGrid" Path="Bounds"/>
    <Binding ElementName="GBHeaderText" Path="Bounds"/>
  </MultiBinding>
</Border.Clip>
```
```
    public class RectGeometryConverter : IMultiValueConverter
    {
        public static readonly RectGeometryConverter Instance = new();

        public object Convert(IList<object> values, Type targetType, object parameter, CultureInfo culture)
        {
            if (values.Count == 2 && values[0] is Rect borderBounds && values[1] is Rect headerBounds)
            {
                return new CombinedGeometry(GeometryCombineMode.Exclude, new RectangleGeometry(borderBounds), new RectangleGeometry(headerBounds));
            }
            return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
        }
    }
```

**Comment by postrondev** [comment]
i am absulute newbee.
do have a small complete example how to use it in a control dll?
thx.

**Comment by BinToss** [comment]
I do have a sample project, but I'm still working out the kinks. In
BinToss/GroupBox.Avalonia, I'm trying to use control themes instead of
styles and one been running into issues like borders being consistently cut
off.

**Comment by BinToss** [comment]
https://github.com/BinToss/GroupBox.Avalonia/releases/v1.0.0
Thanks to @yll690 for the MultiBinding advice!
@postrondev [`BinToss.GroupBox.Avalonia`](https://www.nuget.org/packages/BinToss.GroupBox.Avalonia/) 1.0.0 has been released for Avalonia 11.0.10.
See `./GroupBox.Avalonia.Sample/` for usage.
